onboard/main.cpp

```diff
@@ -0,0 +1,269 @@
+  #include "WProgram.h"
+  #include <TimerThree.h> // Timer library
+  #include "slave_bridge.hpp" //LCM slave file
+  #include "biped_lcm/commData2Teensy.hpp" // header file for data from dispatcher to teensy
+  #include "biped_lcm/commDataFromTeensy.hpp" // header file for data from teensy to dispatcher
+  #include "biped_lcm/error_channel.hpp"
+  #include "Joint.hpp" // struct that stores joint data
+  #include <vector>
+
+  using namespace biped_lcm;
+
+  LCMSerialSlave lcm; //initialize LCM object
+  std::vector<Joint> joints; //vector of joints
+
+  // Channels
+  const int IN  = 0;
+  const int OUT = 1;
+  const int ERROR = 2;
+
+  // PID Variables
+  const int PIDPeriod = 10; // milliseconds
+
+  //State Machine Variables
+  enum STATES{
+    CALIBRATION,
+    RECEIVE_TRAJECTORY,
+    RUN_STATIC_CONTROL,
+    RUN_STATIC_ALL,
+    RUN_TRAJECTORY,
+    RUN_ALL_TRAJECTORIES,
+    WAIT
+  };
+  int state = STATES::WAIT;
+  int currLocalJoint = 0;
+  volatile unsigned long int timer = 0;
+
+  void stopMotors(){
+    for (int i=0; i<numOfJoints; i++){
+      joints[i].motorPWM(zeroTorque);
+      joints[i].setEnable(LOW);
+    }
+  }
+
+  //State machine functions
+  void ID_Request(){
+    commDataFromTeensy msgOut;
+    int32_t links[numOfJoints];
+    for (int i = 0; i<numOfJoints; i++){
+      links[i] = EEPROM.read(joints[i].getMemoryAddr().jointAddr);
+      msgOut.joints[i] = links[i];
+    }
+    lcm.publish(OUT, &msgOut);
+  }
+
+  void Calibration_State(){
+    if ((&joints[currLocalJoint])->CalibrationCheck() == false) {
+      error_channel error_msg;
+      error_msg.potHardwareOutOfRange = true;
+      lcm.publish(ERROR, &error_msg);
+    }
+  }
+
+  void Static_Control_State(){
+    bool OutOfRange = joints[currLocalJoint].checkOOR();
+    if (OutOfRange) {
+      error_channel error_msg;
+      error_msg.isOutOfRange = true;
+      lcm.publish(ERROR, &error_msg);
+      state = STATES::WAIT;
+    }
+    else{
+      (&joints[currLocalJoint])->PIDcontrol();
+    }
+  }
+
+//Convert from Joint number 1-12 to 0-2 for local uses
+  int convertToLocalJointNumber(int num){
+    return ((num-1)%numOfJoints);
+  }
+
+  void Static_Control_All_State(){
+    bool OutOfRange = joints[0].checkOOR()||joints[1].checkOOR()||joints[2].checkOOR();
+    if (OutOfRange) {
+      error_channel error_msg;
+      error_msg.isOutOfRange = true;
+      lcm.publish(ERROR, &error_msg);
+      state = STATES::WAIT;
+    }
+    else{
+      for (int i=0; i<numOfJoints; i++){
+        (&joints[i])->PIDcontrol();
+      }
+    }
+  }
```

Comment:
There is a lot of state machine logic in this file. It might make sense to move it to a seperate file for readability